src/OpenLoco/src/Map/TileManager.cpp

```diff
@@ -480,37 +490,515 @@ namespace OpenLoco::World::TileManager
+            {
+                GameCommands::setErrorText(StringIds::another_company_is_about_to_build_here);
+                return ClearResult::collisionWithErrorMessage;
+            }
+            // Modification from vanilla
+            company->challengeFlags |= CompanyFlags::unk1;
+            return ClearResult::noCollision;
+        };
+
+        if (elSurface == nullptr)
+        {
+            CompanyId owner = getTileOwner(el);
+            if (owner == CompanyId::null)
+            {
+                return ClearResult::noCollision;
+            }
+
+            return returnFunc(owner);
+        }
+        else
+        {
+            auto* elp = &el;
+            while (!elp->isLast())
+            {
+                elp = elp->next();
+                if (!elp->isFlag5())
+                {
+                    continue;
+                }
+                const auto owner = getTileOwner(*elp);
+                if (owner == CompanyId::null)
+                {
+                    continue;
+                }
+
+                return returnFunc(owner);
+            }
+            GameCommands::setErrorText(StringIds::another_company_is_about_to_build_here);
+            return ClearResult::collisionWithErrorMessage;
+        }
+    }
+
+    enum class Sub462B4FResult
+    {
+        allCollisionsRemoved,
+        collision,
+        collisionErrorSet,
+        noCollision,
+        collisionRemoved,
+    };
+
+    // 0x00462937
+    static bool canConstructAtWithClear(const World::Pos2& pos, uint8_t baseZ, uint8_t clearZ, const QuarterTile& qt, uint8_t flags, std::function<ClearResult(const TileElement& el)> clearFunc)
+    {
+        if (!drawableCoords(pos))
+        {
+            GameCommands::setErrorText(StringIds::off_edge_of_map);
+            return false;
+        }
+
+        auto sub_462B4F = [&clearFunc](const TileElement& el) -> Sub462B4FResult {
+            if (clearFunc)
+            {
+                _F0015C = nullptr;
+                if (clearFunc(el) == ClearResult::noCollision)
+                {
+                    if (_F0015C == nullptr)
+                    {
+                        return Sub462B4FResult::noCollision;
+                    }
+                    if (_F0015C == reinterpret_cast<TileElement*>(-1))
+                    {
+                        // This is a no collision as it has removed all the collisions
+                        return Sub462B4FResult::allCollisionsRemoved;
+                    }
+                    // reset tile iterator to _F0015C (its saying we have removed a tile so pointer is stale)
+                    return Sub462B4FResult::collisionRemoved;
+                }
+            }
+            // some stuff getCollideDetails()
+            return Sub462B4FResult::collision;
+        };
+
+        auto checkSurfaceElement = [&clearFunc, &sub_462B4F, clearZ, baseZ, qt](const TileElement& el, const SurfaceElement& elSurface) -> Sub462B4FResult {
```

Comment:
Given the length of this lambda's body, I think it would be more appropriate to extract it to a function of its own. That way, the `canConstructAtWithClear` function can focus on only declaring its business logic.